homeassistant/components/jewish_calendar/sensor.py

```diff
@@ -92,45 +83,101 @@ def make_zmanim(date):
         # tomorrow based on sunset ("shkia"), for others based on "tzais".
         # Hence the following variables.
         after_tzais_date = after_shkia_date = date
-        today_times = make_zmanim(today)
+        today_times = self.make_zmanim(today)
 
         if now > sunset:
             after_shkia_date = date.next_day
 
         if today_times.havdalah and now > today_times.havdalah:
             after_tzais_date = date.next_day
 
+        self._state = self.get_state(after_shkia_date, after_tzais_date)
+        _LOGGER.debug("New value: %s", self._state)
+
+    def make_zmanim(self, date):
+        """Create a Zmanim object."""
+        return hdate.Zmanim(
+            date=date,
+            location=self._location,
+            candle_lighting_offset=self._candle_lighting_offset,
+            havdalah_offset=self._havdalah_offset,
+            hebrew=self._hebrew,
+        )
+
+    def get_state(self, after_shkia_date, after_tzais_date):
+        """For a given type of sensor, return the state."""
         # Terminology note: by convention in py-libhdate library, "upcoming"
         # refers to "current" or "upcoming" dates.
         if self._type == "date":
-            self._state = after_shkia_date.hebrew_date
-        elif self._type == "weekly_portion":
+            return after_shkia_date.hebrew_date
+        if self._type == "weekly_portion":
             # Compute the weekly portion based on the upcoming shabbat.
-            self._state = after_tzais_date.upcoming_shabbat.parasha
-        elif self._type == "holiday_name":
-            self._state = after_shkia_date.holiday_description
-        elif self._type == "holiday_type":
-            self._state = after_shkia_date.holiday_type
-        elif self._type == "upcoming_shabbat_candle_lighting":
-            times = make_zmanim(after_tzais_date.upcoming_shabbat.previous_day.gdate)
-            self._state = times.candle_lighting
-        elif self._type == "upcoming_candle_lighting":
-            times = make_zmanim(
+            return after_tzais_date.upcoming_shabbat.parasha
+        if self._type == "holiday_name":
+            return after_shkia_date.holiday_description
+        if self._type == "holiday_type":
+            return after_shkia_date.holiday_type
+        if self._type == "omer_count":
+            return after_shkia_date.omer_day
+
+        return None
+
+
+class JewishCalendarTimeSensor(JewishCalendarSensor):
+    """Implement attrbutes for sensors returning times."""
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return (
+            self._state.strftime(self._time_format)
```

Comment:
Sensors with timestamp as device class must report an isoformatted absolute utc timestamp as state. That doesn't seem the case here, right?

https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes

Reply:
ISO 8601 doesn't relate to UTC. But, you're right the return value depends on the configuration.

I'll remove the device_class

Reply:
@MartinHjelmare pushed a fix

Reply:
@MartinHjelmare, I'm actually not completely convinced this shouldn't be a timestamp device class: By default I show %H:%M, but the datetime object which defined the state returns all of the data necessary to show it as a full timestamp.

Maybe the key is wrong as it defines a representation, so that should probably be part of Lovelace. And then I return a full timestamp (localized but ISO8601 compliant), but then I'd like a way to hint to Lovelace that the default representation should be %H:%M.

By the way I think the test errored due to something flaky.

Thanks,

Tsvi

Reply:
We should keep the timestamp device class and make sure the state is isoformatted utc. The frontend representation is a frontend matter.

Currently only the user have some options to change the frontend formatting:
https://www.home-assistant.io/lovelace/entities/#format

Reply:
Home Assistant can represent a timestamp as just HH:MM: https://www.home-assistant.io/lovelace/entities/#format

Reply:
@balloob it looks like there is no option for HH:MM. If the use chooses 'time' I suppose it will render with the seconds as well?

Also, is there a way the entity can hint to Lovelace what the default should be? Obviously if the user wants to represent the sensor differently that's their issue.

My reasoning is that I don't want the user to rely on the seconds precision, so I don't want to show it by default.

Reply:
There is currently no way to provide a hint to Lovelace. 

The advantage of using the timestamp device class is that users could have a countdown to the time. I am however not familiar enough to know if that is relevant in this case. Or would everyone always just want to see HH:MM ?

Reply:
It definitely is relevant. Could be a nice touch for a countdown towards candle lighting especially.

I reverted the removal of the timestamp device class. The state is a datetime which is timezone aware, so it should show up correctly (or should I call `self._state.isoformat() if not None` and removed the time format configuration, as it should be solved in the frontend.

I'm wondering if we should add the attributes I added here to all device_class == "timestamp"? Could have a nice use case for working with input_datetime/sensor.time?

Reply:
The timestamp should be UTC, the browser will convert it to the relevant local time.

Reply:
Won't that cause issues with automations trying to compare to `sensor.time`? Or how else would I check if a given time has arrived?

Reply:
`utcnow()` is available in templates in automations.

Reply:
@balloob, @MartinHjelmare I checked the UI, it doesn't change if the state is in UTC or not. As long as it's not a datetime naive result the UI will show it correctly.
It would be better to leave it as non-UTC as it makes more sense when looking at the data. (When clicking on the sensor it looks weird seeing the time in UTC while the time in the general UI is local, you could compare it to looking at a timestamp for the next bus arrival).
So unless I hear a strong case for changing it to UTC, I'm inclined to leave it localized.

ISO8601 does *not* specify UTC timezone, rather it specifies the format *how* the timezone data is to be represented.

Reply:
All our times in the state and state attributes should be utc. That makes it easy to reason about and users know what to expect.

The frontend is a separate concern.

Reply:
Ok, so in the case of a timestamp showing the bus arrival, the state should also UTC time?

Reply:
Yes.